Couldn't find language#92 in [repository_id#8]

The following query:

```sql
SELECT
                    language,
                    COUNT(repository_id) AS repository_count
                FROM (
                    SELECT DISTINCT
                        repository_id,
                        COALESCE(language(file_path, blob_content), 'Unknown') AS language
                    FROM ref_commits
                    NATURAL JOIN commit_files
                    NATURAL JOIN files
                    WHERE
                        ref_name LIKE 'refs/heads/HEAD/%'
                    ) AS q2
                GROUP BY language
                ORDER BY repository_count DESC
```

produces the following error:

```
Exchange hashpartitioning(language#92, 200)
+- *(2) HashAggregate(keys=[language#92], functions=[partial_count(repository_id#8)], output=[language#92, count#109L])
   +- *(2) HashAggregate(keys=[repository_id#8, language#92], functions=[], output=[repository_id#8, language#92])
      +- Exchange hashpartitioning(repository_id#8, language#92, 200)
         +- *(1) HashAggregate(keys=[repository_id#8, language#92], functions=[], output=[repository_id#8, language#92])
            +- *(1) DataSourceV2Scan [repository_id#8], DefaultReader(List(GitbaseServer(localhost:3308,root,)),StructType(StructField(repository_id,StringType,true), StructField(UDF:language(file_path#52, cast(blob_content#83 as binary)),StringType,true), StructField(Unknown,StringType,false)),Project(List(repository_id#8, UDF:language(file_path#52, cast(blob_content#83 as binary)) AS UDF:language(file_path#52, cast(blob_content#83 as binary))#105, Unknown AS Unknown#106),Filter(List(StartsWith(ref_name#10, refs/heads/HEAD/)),Join(Join(Table(ref_commits),Table(commit_files),Some(((repository_id#8 = repository_id#50) && (commit_hash#9 = commit_hash#51)))),Table(files),Some(((((repository_id#8 = repository_id#78) && (file_path#52 = file_path#79)) && (blob_hash#53 = blob_hash#80)) && (tree_hash#54 = tree_hash#81)))))))

org.apache.spark.sql.catalyst.errors.package$TreeNodeException: execute, tree:
Exchange hashpartitioning(language#92, 200)
+- *(2) HashAggregate(keys=[language#92], functions=[partial_count(repository_id#8)], output=[language#92, count#109L])
   +- *(2) HashAggregate(keys=[repository_id#8, language#92], functions=[], output=[repository_id#8, language#92])
      +- Exchange hashpartitioning(repository_id#8, language#92, 200)
         +- *(1) HashAggregate(keys=[repository_id#8, language#92], functions=[], output=[repository_id#8, language#92])
            +- *(1) DataSourceV2Scan [repository_id#8], DefaultReader(List(GitbaseServer(localhost:3308,root,)),StructType(StructField(repository_id,StringType,true), StructField(UDF:language(file_path#52, cast(blob_content#83 as binary)),StringType,true), StructField(Unknown,StringType,false)),Project(List(repository_id#8, UDF:language(file_path#52, cast(blob_content#83 as binary)) AS UDF:language(file_path#52, cast(blob_content#83 as binary))#105, Unknown AS Unknown#106),Filter(List(StartsWith(ref_name#10, refs/heads/HEAD/)),Join(Join(Table(ref_commits),Table(commit_files),Some(((repository_id#8 = repository_id#50) && (commit_hash#9 = commit_hash#51)))),Table(files),Some(((((repository_id#8 = repository_id#78) && (file_path#52 = file_path#79)) && (blob_hash#53 = blob_hash#80)) && (tree_hash#54 = tree_hash#81)))))))

	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:56)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec.doExecute(ShuffleExchangeExec.scala:119)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.InputAdapter.inputRDDs(WholeStageCodegenExec.scala:371)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.inputRDDs(HashAggregateExec.scala:150)
	at org.apache.spark.sql.execution.WholeStageCodegenExec.doExecute(WholeStageCodegenExec.scala:605)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.TakeOrderedAndProjectExec.executeCollect(limit.scala:135)
	at org.apache.spark.sql.Dataset.org$apache$spark$sql$Dataset$$collectFromPlan(Dataset.scala:3273)
	at org.apache.spark.sql.Dataset$$anonfun$head$1.apply(Dataset.scala:2484)
	at org.apache.spark.sql.Dataset$$anonfun$head$1.apply(Dataset.scala:2484)
	at org.apache.spark.sql.Dataset$$anonfun$52.apply(Dataset.scala:3254)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:77)
	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3253)
	at org.apache.spark.sql.Dataset.head(Dataset.scala:2484)
	at org.apache.spark.sql.Dataset.take(Dataset.scala:2698)
	at org.apache.spark.sql.Dataset.showString(Dataset.scala:254)
	at org.apache.spark.sql.Dataset.show(Dataset.scala:725)
	at org.apache.spark.sql.Dataset.show(Dataset.scala:702)
	at tech.sourced.gitbase.spark.DefaultSourceSpec$$anonfun$14.apply$mcV$sp(DefaultSourceSpec.scala:397)
	at tech.sourced.gitbase.spark.DefaultSourceSpec$$anonfun$14.apply(DefaultSourceSpec.scala:397)
	at tech.sourced.gitbase.spark.DefaultSourceSpec$$anonfun$14.apply(DefaultSourceSpec.scala:397)
	at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FlatSpecLike$$anon$1.apply(FlatSpecLike.scala:1682)
	at org.scalatest.TestSuite$class.withFixture(TestSuite.scala:196)
	at org.scalatest.FlatSpec.withFixture(FlatSpec.scala:1685)
	at org.scalatest.FlatSpecLike$class.invokeWithFixture$1(FlatSpecLike.scala:1679)
	at org.scalatest.FlatSpecLike$$anonfun$runTest$1.apply(FlatSpecLike.scala:1692)
	at org.scalatest.FlatSpecLike$$anonfun$runTest$1.apply(FlatSpecLike.scala:1692)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
	at org.scalatest.FlatSpecLike$class.runTest(FlatSpecLike.scala:1692)
	at org.scalatest.FlatSpec.runTest(FlatSpec.scala:1685)
	at org.scalatest.FlatSpecLike$$anonfun$runTests$1.apply(FlatSpecLike.scala:1750)
	at org.scalatest.FlatSpecLike$$anonfun$runTests$1.apply(FlatSpecLike.scala:1750)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:396)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
	at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:373)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:410)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
	at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:379)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
	at org.scalatest.FlatSpecLike$class.runTests(FlatSpecLike.scala:1750)
	at org.scalatest.FlatSpec.runTests(FlatSpec.scala:1685)
	at org.scalatest.Suite$class.run(Suite.scala:1147)
	at org.scalatest.FlatSpec.org$scalatest$FlatSpecLike$$super$run(FlatSpec.scala:1685)
	at org.scalatest.FlatSpecLike$$anonfun$run$1.apply(FlatSpecLike.scala:1795)
	at org.scalatest.FlatSpecLike$$anonfun$run$1.apply(FlatSpecLike.scala:1795)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
	at org.scalatest.FlatSpecLike$class.run(FlatSpecLike.scala:1795)
	at tech.sourced.gitbase.spark.DefaultSourceSpec.org$scalatest$BeforeAndAfterAll$$super$run(DefaultSourceSpec.scala:4)
	at org.scalatest.BeforeAndAfterAll$class.liftedTree1$1(BeforeAndAfterAll.scala:213)
	at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:210)
	at tech.sourced.gitbase.spark.DefaultSourceSpec.run(DefaultSourceSpec.scala:4)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:45)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$1.apply(Runner.scala:1346)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$1.apply(Runner.scala:1340)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:1340)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1011)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1010)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:1506)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1010)
	at org.scalatest.tools.Runner$.run(Runner.scala:850)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2(ScalaTestRunner.java:131)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:28)
Caused by: org.apache.spark.sql.catalyst.errors.package$TreeNodeException: execute, tree:
Exchange hashpartitioning(repository_id#8, language#92, 200)
+- *(1) HashAggregate(keys=[repository_id#8, language#92], functions=[], output=[repository_id#8, language#92])
   +- *(1) DataSourceV2Scan [repository_id#8], DefaultReader(List(GitbaseServer(localhost:3308,root,)),StructType(StructField(repository_id,StringType,true), StructField(UDF:language(file_path#52, cast(blob_content#83 as binary)),StringType,true), StructField(Unknown,StringType,false)),Project(List(repository_id#8, UDF:language(file_path#52, cast(blob_content#83 as binary)) AS UDF:language(file_path#52, cast(blob_content#83 as binary))#105, Unknown AS Unknown#106),Filter(List(StartsWith(ref_name#10, refs/heads/HEAD/)),Join(Join(Table(ref_commits),Table(commit_files),Some(((repository_id#8 = repository_id#50) && (commit_hash#9 = commit_hash#51)))),Table(files),Some(((((repository_id#8 = repository_id#78) && (file_path#52 = file_path#79)) && (blob_hash#53 = blob_hash#80)) && (tree_hash#54 = tree_hash#81)))))))

	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:56)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec.doExecute(ShuffleExchangeExec.scala:119)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.InputAdapter.inputRDDs(WholeStageCodegenExec.scala:371)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.inputRDDs(HashAggregateExec.scala:150)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.inputRDDs(HashAggregateExec.scala:150)
	at org.apache.spark.sql.execution.WholeStageCodegenExec.doExecute(WholeStageCodegenExec.scala:605)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec.prepareShuffleDependency(ShuffleExchangeExec.scala:92)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec$$anonfun$doExecute$1.apply(ShuffleExchangeExec.scala:128)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec$$anonfun$doExecute$1.apply(ShuffleExchangeExec.scala:119)
	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:52)
	... 82 more
Caused by: org.apache.spark.sql.catalyst.errors.package$TreeNodeException: Binding attribute, tree: language#92
	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:56)
	at org.apache.spark.sql.catalyst.expressions.BindReferences$$anonfun$bindReference$1.applyOrElse(BoundAttribute.scala:91)
	at org.apache.spark.sql.catalyst.expressions.BindReferences$$anonfun$bindReference$1.applyOrElse(BoundAttribute.scala:90)
	at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$2.apply(TreeNode.scala:267)
	at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$2.apply(TreeNode.scala:267)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:70)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:266)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transform(TreeNode.scala:256)
	at org.apache.spark.sql.catalyst.expressions.BindReferences$.bindReference(BoundAttribute.scala:90)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec$$anonfun$48.apply(HashAggregateExec.scala:737)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec$$anonfun$48.apply(HashAggregateExec.scala:737)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.immutable.List.map(List.scala:296)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.doConsumeWithKeys(HashAggregateExec.scala:737)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.doConsume(HashAggregateExec.scala:173)
	at org.apache.spark.sql.execution.CodegenSupport$class.constructDoConsumeFunction(WholeStageCodegenExec.scala:208)
	at org.apache.spark.sql.execution.CodegenSupport$class.consume(WholeStageCodegenExec.scala:179)
	at org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanExec.consume(DataSourceV2ScanExec.scala:37)
	at org.apache.spark.sql.execution.ColumnarBatchScan$class.produceRows(ColumnarBatchScan.scala:166)
	at org.apache.spark.sql.execution.ColumnarBatchScan$class.doProduce(ColumnarBatchScan.scala:80)
	at org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanExec.doProduce(DataSourceV2ScanExec.scala:37)
	at org.apache.spark.sql.execution.CodegenSupport$$anonfun$produce$1.apply(WholeStageCodegenExec.scala:88)
	at org.apache.spark.sql.execution.CodegenSupport$$anonfun$produce$1.apply(WholeStageCodegenExec.scala:83)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.CodegenSupport$class.produce(WholeStageCodegenExec.scala:83)
	at org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanExec.produce(DataSourceV2ScanExec.scala:37)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.doProduceWithKeys(HashAggregateExec.scala:641)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.doProduce(HashAggregateExec.scala:165)
	at org.apache.spark.sql.execution.CodegenSupport$$anonfun$produce$1.apply(WholeStageCodegenExec.scala:88)
	at org.apache.spark.sql.execution.CodegenSupport$$anonfun$produce$1.apply(WholeStageCodegenExec.scala:83)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.CodegenSupport$class.produce(WholeStageCodegenExec.scala:83)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.produce(HashAggregateExec.scala:39)
	at org.apache.spark.sql.execution.WholeStageCodegenExec.doCodeGen(WholeStageCodegenExec.scala:524)
	at org.apache.spark.sql.execution.WholeStageCodegenExec.doExecute(WholeStageCodegenExec.scala:576)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec.prepareShuffleDependency(ShuffleExchangeExec.scala:92)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec$$anonfun$doExecute$1.apply(ShuffleExchangeExec.scala:128)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec$$anonfun$doExecute$1.apply(ShuffleExchangeExec.scala:119)
	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:52)
	... 103 more
Caused by: java.lang.RuntimeException: Couldn't find language#92 in [repository_id#8]
	at scala.sys.package$.error(package.scala:27)
	at org.apache.spark.sql.catalyst.expressions.BindReferences$$anonfun$bindReference$1$$anonfun$applyOrElse$1.apply(BoundAttribute.scala:97)
	at org.apache.spark.sql.catalyst.expressions.BindReferences$$anonfun$bindReference$1$$anonfun$applyOrElse$1.apply(BoundAttribute.scala:91)
	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:52)
	... 154 more
```

--- a/src/main/scala/tech/sourced/gitbase/spark/rule/PushdownTree.scala
+++ b/src/main/scala/tech/sourced/gitbase/spark/rule/PushdownTree.scala
@@ -137,47 +137,56 @@ object PushdownTree extends Rule[LogicalPlan] {
     * Split expressions in supported and unsupported to push down
     * only the supported ones and keep upside the unsupported
     */
-  private def splitSupportedAndUnsupportedExpressions(expressions: Seq[NamedExpression]):
+  private def splitSupportedAndUnsupportedExpressions(expressions: Seq[Expression]):
   (Seq[NamedExpression], Seq[NamedExpression]) = {
-    val (s, u) = expressions.map { exp =>
-      var supported: Seq[NamedExpression] = Nil
-      var unsupported: Option[NamedExpression] = None
-      exp.foreach {
-        // this expression cannot be handled but its children can
-        case e if !canBeHandled(e :: Nil) && canBeHandled(e.children) && unsupported.isEmpty =>
-          val (supportedExp, supportedReference) =
-            e.children.map { e =>
-              val namedExpression = toNamedExpression(e)
-              (namedExpression,
-                AttributeReference(
-                  namedExpression.name,
-                  namedExpression.dataType,
-                  namedExpression.nullable,
-                  namedExpression.metadata)
-                (namedExpression.exprId,
-                  namedExpression.qualifier)
-              )
-            }.unzip
+    import scala.collection.mutable
+    val unsupported = mutable.Buffer[NamedExpression]()
+    val supported = mutable.Buffer[NamedExpression]()
 
-          val unsupportedNamedExp = toNamedExpression(e.withNewChildren(supportedReference))
+    expressions.foreach(exp => {
+      if (canBeHandled(exp :: Nil)) {
+        val e = toNamedExpression(exp)
+        supported += e
+        unsupported += AttributeReference(
+          e.name,
+          e.dataType,
+          e.nullable,
+          e.metadata
+        )(e.exprId, e.qualifier)
+      } else if (!canBeHandled(exp :: Nil) && canBeHandled(exp.children)) {
+        unsupported += toNamedExpression(exp.withNewChildren(exp.children.map(c => {
+          val child = toNamedExpression(c)
+          supported += child
+          AttributeReference(
+            child.name,
+            child.dataType,
+            child.nullable,
+            child.metadata
+          )(child.exprId, child.qualifier)
+        })))
+      } else if (exp.children.nonEmpty) {
+        unsupported += toNamedExpression(exp.withNewChildren(exp.children.map(c => {
+          val (s, u) = splitSupportedAndUnsupportedExpressions(c :: Nil)
+          if (u.length != 1) {
+            throw new SparkException("invalid number of unsupported expressions obtained")
+          }
 
-          supported = supportedExp
-          unsupported = Some(unsupportedNamedExp)
-        // this expression can be handled and it has no children
-        case e if canBeHandled(e :: Nil) && unsupported.isEmpty =>
-          val namedExp = toNamedExpression(e)
-          supported = Seq(namedExp)
-          unsupported = Some(
-            AttributeReference(
-              namedExp.name, namedExp.dataType, namedExp.nullable, namedExp.metadata)
-            (namedExp.exprId, namedExp.qualifier))
-        case _ =>
+          supported ++= s
+          u.head match {
+            case e: Alias => e.child // remove aliases from non topmost expressions
+            case o => o
+          }
+        })))
+      } else {
+        unsupported += toNamedExpression(exp)
       }
+    })
 
-      (supported, unsupported
-        .getOrElse(throw new SparkException("No unsupported expressions on the query")))
-    }.unzip
+    if (unsupported.length != expressions.length) {
+      throw new SparkException("unsupported exceptions don't match input expressions")
+    }
 
-    (s.flatten, u)
+    (supported.toList, unsupported.toList)
   }
+
 }

--- a/src/test/scala/tech/sourced/gitbase/spark/DefaultSourceSpec.scala
+++ b/src/test/scala/tech/sourced/gitbase/spark/DefaultSourceSpec.scala
@@ -210,11 +210,11 @@ class DefaultSourceSpec extends BaseGitbaseSpec {
   }
 
   it should "get commit parents using parse_commit_parents" in {
-    val result = spark.sql("SELECT repository_id, parse_commit_parents(commit_parents)" +
+    val result = spark.sql("SELECT repository_id, parse_commit_parents(commit_parents) AS parents" +
       " FROM ref_commits" +
       " NATURAL JOIN commits" +
       " WHERE ref_name LIKE 'refs/heads/HEAD/%' AND history_index = 0" +
-      " ORDER BY repository_id, 2")
+      " ORDER BY repository_id, parents")
       .collect()
       .map(row => (row(0), row(1).asInstanceOf[Seq[String]]))
 
@@ -279,4 +279,5 @@ class DefaultSourceSpec extends BaseGitbaseSpec {
       ("fff840f8784ef162dc83a1465fc5763d890b68ba", 2)
     ))
   }
+
 }

